Custom authentication (#12)
victors1681/dgii-ecf: README.md

[More Information ARECF](https://dgii.gov.do/cicloContribuyente/facturacion/comprobantesFiscalesElectronicosE-CF/Documentacin%20sobre%20eCF/Formatos%20XML/Formato%20Acuse%20de%20Recibo%20v%201.0.pdf)

## Custom Authentication

The `CustomAuthentication` class provides methods to create a custom authentication, generate a seed, validate a seed, verify XML signatures, as well as generate and verify JWT tokens.

### Usage

#### Step One | Initialize the class

Initialize the `CustomAuthentication` class with the necessary certificate.

```typescript
import { CustomAuthentication, P12Reader } from 'dgii-ecf';

const secret = process.env.CERTIFICATE_TEST_PASSWORD || '';

const reader = new P12Reader(secret);
const certs = reader.getKeyFromFile(
  path.resolve(__dirname, `../../test_cert/'your certificate'}`)
);

const customAuth = new CustomAuthentication(cert);
```

#### Step Two | Generate and send the seed

> Generate a seed and send it to the client. Ensure the client accesses the seed via the endpoint following the DGII standard:

Reviewer: ```suggestion
Generate a seed and send it to the client. Ensure the client can access the seed via the endpoint following the DGII standard:
```

`{your_host}/fe/autenticacion/api/semilla`

```ts

> const seed = customAuthentication.generateSeed();

Reviewer: ```suggestion
const seed = customAuth.generateSeed();
```

_Following the initialization from step one_

```

#### Step Three | Validate the Seed

Create an endpoint to validate the seed submitted by the client. The endpoint URL should follow the DGII standard:
`{your_host}/fe/autenticacion/api/validacioncertificado`

```ts

> const token = await customAuthentication.verifySignedSeed(signedSeed);

Reviewer: ```suggestion
const token = await customAuth.verifySignedSeed(seed);
```

```

#### Step Four | Validate the Seed

Once the authentication flow is complete, restrict the client’s access to your resources. Validate the access token before allowing access to the protected endpoint like `/fe/recepción/api/ecf`

```ts
try {

>   const decoded = await customAuthentication.verifyToken(token);

Reviewer: ```suggestion
  const decoded = await customAuth.verifyToken(token);
```

} catch (error) {
  console.error('Token verification failed:', error.message);
}
```

## Run Integration Test local environment

This repo performs the unit/integration tests connecting to the DGII test environment

> - First: In order to pass the test locally the first step is to plate your certificate into the directory `src/test_cert` current emtpy, for security reason `.p12` gets ignored.

Reviewer: ```suggestion
- First: In order to pass the test locally the first step is to place your certificate into the directory `src/test_cert` current emtpy, for security reason `.p12` gets ignored.
```
